File: src/model/schema.ts
// The normalized diagram model. Authoring YAML is parsed and validated into
// these shapes with every default filled in, so downstream stages (layout,
// route, render, animate) never deal with optional/raw input.

/** What the diagram *is*. Every document declares one via a root `type:` key
 *  (`architecture` is the layered node/edge graph Beck launched with). The type
 *  picks the layout + routing strategy; measure/render/animate are shared. */
export type DiagramType = 'architecture' | 'sequence' | 'state' | 'class'
export type Direction = 'TB' | 'BT' | 'LR' | 'RL'
export type ThemeMode = 'auto' | 'light' | 'dark'
/** How a diagram wider than its container behaves. `shrink` (default) scales the
 *  whole diagram down to fit the available width; `scroll` keeps it at natural
 *  size and lets the container scroll horizontally. Vertical size is unaffected. */
export type FitMode = 'shrink' | 'scroll'
export type NodeKind =
  | 'service'
  | 'db'
  | 'queue'
  | 'cache'
  | 'gateway'
  | 'external'
  | 'user'
  | 'ghost'
export type NodeVariant = 'solid' | 'subtle' | 'ghost'
/** Structural form of a node. `card` is the classic architecture card; `pill`
 *  is a state-diagram state; `start`/`end` are the state machine entry/exit
 *  dots; `class` is a multi-compartment UML class card. */
export type NodeShape = 'card' | 'pill' | 'start' | 'end' | 'class'
export type EdgeStyle = 'solid' | 'dashed'
export type EdgeCurve = 'step-round' | 'straight' | 's'
export type EdgeKind = 'data' | 'control' | 'async' | 'dependency'
/** Easing token for a travelling packet (mapped to a concrete GSAP ease in the animator). */
export type PacketEase =
  | 'linear'
  | 'smooth'
  | 'accelerate'
  | 'decelerate'
  | 'expo'
  | 'sine'
  | 'steps'
  | 'bounce'
/** Visual form of a travelling packet. `dot` is the default small glowing dot;
 *  `circle` is a larger filled disc; `ring` is a hollow stroked circle. */
export type PacketShape = 'dot' | 'circle' | 'ring'
export type Side = 'top' | 'bottom' | 'left' | 'right'
/** Which ends of an edge carry an arrowhead. */
export type ArrowEnds = 'none' | 'end' | 'start' | 'both'
/** End decoration on an edge. Beyond the classic filled `arrow`: `arrow-open`
 *  (a stroked chevron — async/reply messages, dependencies), `triangle`
 *  (hollow — UML inheritance), and `diamond`/`diamond-open` (UML composition/
 *  aggregation). When set on an edge these win over the `arrow` ends. */
export type MarkerShape = 'arrow' | 'arrow-open' | 'triangle' | 'diamond' | 'diamond-open'
export type AccentToken = 'primary' | 'success' | 'warn' | 'danger' | 'info' | 'neutral'

export interface Spacing {
  rank: number
  node: number
  cornerRadius: number
}

export interface DiagramMeta {
  type: DiagramType
  title?: string
  subtitle?: string
  direction: Direction
  theme: ThemeMode
  animate: boolean
  loop: boolean
  /** How the diagram behaves when wider than its container (`shrink`/`scroll`). */
  fit: FitMode
  spacing: Spacing
}

export interface NodeModel {
  id: string
  title: string
  subtitle?: string
  /** Named icon key OR raw inline `<svg>` markup. Resolved at render time. */
  icon?: string
  kind: NodeKind
  variant: NodeVariant
  status?: string
  /** CSS color value, e.g. `var(--beck-primary)` or a raw hex. */
  accent: string
  /** Optional link target; renders the card as an `<a href>`. */
  href?: string
  /** Anchor target (e.g. `_blank`); only meaningful with `href`. */
  target?: string
  /** Override the card background (CSS color); defaults to the theme surface. */
  surface?: string
  /** Override the card text color (CSS color); defaults to the theme text. */
  textColor?: string
  width?: number
  rank?: number
  order?: number
  group?: string
  /** Structural form; `card` for architecture nodes and sequence participants. */
  shape: NodeShape
  /** UML stereotype rendered as «stereotype» above a class name. */
  stereotype?: string
  /** Class-card field compartment lines (empty for non-class shapes). */
  fields: string[]
  /** Class-card method compartment lines (empty for non-class shapes). */
  methods: string[]
}

export interface GroupModel {
  id: string
  label: string
  members: string[]
  accent: string
}

export interface EdgeModel {
  id: string
  from: string
  to: string
  label?: string
  style: EdgeStyle
  curve: EdgeCurve
  kind: EdgeKind
  /** CSS color value for the stroke. */
  color: string
  /** Which ends carry an arrowhead (`true`/`false` authoring maps to end/none). */
  arrow: ArrowEnds
  fromSide?: Side
  toSide?: Side
  /** Explicit end decorations; when set they win over `arrow` at that end. */
  markerStart?: MarkerShape
  markerEnd?: MarkerShape
  /** Small annotation near the source end (UML multiplicity, e.g. "1"). */
  fromLabel?: string
  /** Small annotation near the target end (UML multiplicity, e.g. "*"). */
  toLabel?: string
  /** Sequence only: a dashed return message (rendered with an open arrowhead). */
  reply: boolean
  /** Sequence only: force-start (true) or suppress (false) an activation bar on
   *  the receiver; undefined lets the request/reply heuristic decide. */
  activate?: boolean
}

/** Knobs shared by `packet` and `burst` that shape the travelling dot. All are
 *  optional; unset ones fall back to the edge-kind defaults, then to engine
 *  constants. */
export interface PacketKnobs {
  /** Visual form of the packet (`dot` | `circle` | `ring`); defaults to `dot`. */
  shape?: PacketShape
  /** Dot radius in px. */
  size?: number
  /** Travel speed in px/s. */
  speed?: number
  /** Soft glow on the dot. */
  glow?: boolean
  /** Emit an expanding ring (a "burst") at the destination on arrival. */
  impact?: boolean
  /** Easing of the dot's travel (and the trail draw, kept in lockstep). */
  ease?: PacketEase
}

export type FlowStep =
  /** `edge` pins the step to one specific edge id — used by sequence diagrams,
   *  where many messages share the same from/to pair. */
  | ({ type: 'packet'; from: string; to: string; via?: string[]; edge?: string; color?: string; label?: string } & PacketKnobs)
  /** Several dots down an edge (or fanned to many targets), staggered. */
  | ({
      type: 'burst'
      from: string
      to: string | string[]
      via?: string[]
      count: number
      stagger: number
      color?: string
      label?: string
    } & PacketKnobs)
  | { type: 'status'; node: string; text: string; color?: string }
  | { type: 'highlight'; node: string; color?: string }
  | { type: 'pulse'; node: string; color?: string }
  /** Persistently recolor an edge (and its arrowhead) until the next reset. */
  | { type: 'activate'; from: string; to: string; color?: string }
  /** Continuous flowing dashes along an edge (ongoing traffic), until reset. */
  | { type: 'stream'; from: string; to: string; color?: string }
  /** Leave a node visibly busy (breathing glow) until `idle`/reset. */
  | { type: 'working'; node: string; color?: string }
  /** Clear a node's `working` state. */
  | { type: 'idle'; node: string }
  /** A failure beat: red shake + flash, with an optional status text. */
  | { type: 'fail'; node: string; text?: string; color?: string }
  | { type: 'phase'; label: string }
  | { type: 'wait'; seconds: number }
  | { type: 'reset' }
  | { type: 'parallel'; steps: FlowStep[] }

export interface FlowModel {
  repeat: number
  repeatDelay: number
  steps: FlowStep[]
  /** True when the flow was auto-derived from the edges (no `flow:` authored). */
  derived: boolean
}

/** A labelled horizontal band in a sequence diagram, drawn before message `at`. */
export interface SectionMark {
  label: string
  at: number
}

export interface DiagramModel {
  meta: DiagramMeta
  nodes: NodeModel[]
  groups: GroupModel[]
  edges: EdgeModel[]
  flow: FlowModel
  /** Sequence section bands (empty for other diagram types). */
  sections: SectionMark[]
}

File: dotnet/Beck/Model/Schema.cs
namespace Beck.Rendering;

// The normalized diagram model — a faithful port of src/model/schema.ts. Authoring
// YAML is parsed and validated into these shapes with every default filled in, so
// downstream stages (layout, route, render, animate) never see optional/raw input.
// Types are `internal`: consumers touch only BeckSvg + SvgRenderOptions.

internal sealed record Spacing(double Rank, double Node, double CornerRadius);

/// <summary>Narration caption behaviour + reading-time pacing.</summary>
internal sealed record NarrationOptions(bool Enabled, double Wpm, double Min, double Pad);

internal sealed record DiagramMeta
{
    public required DiagramType Type { get; init; }
    public string? Title { get; init; }
    public string? Subtitle { get; init; }
    /// <summary>The raw <c>meta.style</c> token (<c>[a-z0-9-]+</c>), or null when unset/malformed.
    /// Resolved to a <c>BeckStyle</c> in <c>BeckSvg</c> (resolution needs the render options the
    /// model never sees), mirroring how <see cref="Title"/> stays a plain string here.</summary>
    public string? StyleName { get; init; }
    public required Direction Direction { get; init; }
    public required ThemeMode Theme { get; init; }
    /// <summary>Mutable: the class builder forces this false when no flow is authored.</summary>
    public required bool Animate { get; set; }
    public required bool Loop { get; init; }
    public required FitMode Fit { get; init; }
    public required Spacing Spacing { get; init; }
    public required NarrationOptions Narration { get; init; }
}

internal sealed record NodeModel
{
    public required string Id { get; init; }
    public required string Title { get; init; }
    public string? Subtitle { get; init; }
    /// <summary>Named icon key OR raw inline <c>&lt;svg&gt;</c> markup. Resolved at render time.</summary>
    public string? Icon { get; init; }
    public required NodeKind Kind { get; init; }
    public required NodeVariant Variant { get; init; }
    public string? Status { get; init; }
    /// <summary>CSS color value, e.g. <c>var(--beck-primary)</c> or a raw hex.</summary>
    public required string Accent { get; init; }
    public string? Href { get; init; }
    public string? Target { get; init; }
    public string? Surface { get; init; }
    public string? TextColor { get; init; }
    public double? Width { get; init; }
    public double? Rank { get; init; }
    public double? Order { get; init; }
    public string? Group { get; init; }
    public required NodeShape Shape { get; init; }
    public string? Stereotype { get; init; }
    public required IReadOnlyList<string> Fields { get; init; }
    public required IReadOnlyList<string> Methods { get; init; }
}

internal sealed record GroupModel
{
    public required string Id { get; init; }
    public required string Label { get; set; }
    /// <summary>Node or nested-group ids; populated during group building.</summary>
    public required List<string> Members { get; init; }
    public required string Accent { get; init; }
}

internal sealed record EdgeModel
{
    public required string Id { get; init; }
    public required string From { get; init; }
    public required string To { get; init; }
    public string? Label { get; init; }
    public required EdgeStyle Style { get; init; }
    public required EdgeCurve Curve { get; init; }
    public required EdgeKind Kind { get; init; }
    /// <summary>CSS color value for the stroke.</summary>
    public required string Color { get; init; }
    public required ArrowEnds Arrow { get; init; }
    public string? Note { get; init; }
    public Side? FromSide { get; init; }
    public Side? ToSide { get; init; }
    public MarkerShape? MarkerStart { get; init; }
    public MarkerShape? MarkerEnd { get; init; }
    public string? FromLabel { get; init; }
    public string? ToLabel { get; init; }
    /// <summary>Sequence only: a dashed return message.</summary>
    public required bool Reply { get; init; }
    /// <summary>Sequence only: force/suppress an activation bar; null lets the heuristic decide.</summary>
    public bool? Activate { get; init; }
}

/// <summary>Shared <c>packet</c>/<c>burst</c> motion knobs; unset ones fall back to edge-kind defaults.</summary>
internal sealed record PacketKnobs
{
    public PacketShape? Shape { get; init; }
    public double? Size { get; init; }
    public double? Speed { get; init; }
    public bool? Glow { get; init; }
    public bool? Impact { get; init; }
    public PacketEase? Ease { get; init; }
}

/// <summary>A single flow step. Discriminated subtypes mirror the TS <c>FlowStep</c> union.</summary>
internal abstract record FlowStep
{
    /// <summary>The wire discriminator (<c>packet</c>, <c>burst</c>, …).</summary>
    public abstract string Kind { get; }
}

internal sealed record PacketStep : FlowStep
{
    public override string Kind => "packet";
    public required string From { get; init; }
    public required string To { get; init; }
    public IReadOnlyList<string>? Via { get; init; }
    public string? Edge { get; init; }
    public string? Color { get; init; }
    public string? Label { get; init; }
    public required PacketKnobs Knobs { get; init; }
}

internal sealed record BurstStep : FlowStep
{
    public override string Kind => "burst";
    public required string From { get; init; }
    /// <summary>Single target (mutually exclusive with <see cref="ToList"/>).</summary>
    public string? To { get; init; }
    /// <summary>Fan-out targets when authored as a list.</summary>
    public IReadOnlyList<string>? ToList { get; init; }
    public IReadOnlyList<string>? Via { get; init; }
    public required int Count { get; init; }
    public required double Stagger { get; init; }
    public string? Color { get; init; }
    public string? Label { get; init; }
    public required PacketKnobs Knobs { get; init; }

    /// <summary>The targets as a flat list, however they were authored.</summary>
    public IEnumerable<string> Targets => ToList ?? (To is null ? [] : [To]);
}

internal sealed record StatusStep : FlowStep
{
    public override string Kind => "status";
    public required string Node { get; init; }
    public required string Text { get; init; }
    public string? Color { get; init; }
}

internal sealed record HighlightStep : FlowStep
{
    public override string Kind => "highlight";
    public required string Node { get; init; }
    public string? Color { get; init; }
}

internal sealed record PulseStep : FlowStep
{
    public override string Kind => "pulse";
    public required string Node { get; init; }
    public string? Color { get; init; }
}

internal sealed record ActivateStep : FlowStep
{
    public override string Kind => "activate";
    public required string From { get; init; }
    public required string To { get; init; }
    public string? Color { get; init; }
}

internal sealed record StreamStep : FlowStep
{
    public override string Kind => "stream";
    public required string From { get; init; }
    public required string To { get; init; }
    public string? Color { get; init; }
}

internal sealed record WorkingStep : FlowStep
{
    public override string Kind => "working";
    public required string Node { get; init; }
    public string? Color { get; init; }
}

internal sealed record IdleStep : FlowStep
{
    public override string Kind => "idle";
    public required string Node { get; init; }
}

internal sealed record FailStep : FlowStep
{
    public override string Kind => "fail";
    public required string Node { get; init; }
    public string? Text { get; init; }
    public string? Color { get; init; }
}

internal sealed record NarrateStep : FlowStep
{
    public override string Kind => "narrate";
    public required string Text { get; init; }
    public double? Hold { get; init; }
    public string? Color { get; init; }
}

internal sealed record PhaseStep : FlowStep
{
    public override string Kind => "phase";
    public required string Label { get; init; }
}

internal sealed record WaitStep : FlowStep
{
    public override string Kind => "wait";
    public required double Seconds { get; init; }
}

internal sealed record ResetStep : FlowStep
{
    public override string Kind => "reset";
}

internal sealed record ParallelStep : FlowStep
{
    public override string Kind => "parallel";
    public required IReadOnlyList<FlowStep> Steps { get; init; }
}

internal sealed record FlowModel
{
    /// <summary>-1 infinite / 0 once / N. Mutable: <c>meta.loop:false</c> forces 0.</summary>
    public required double Repeat { get; set; }
    public required double RepeatDelay { get; init; }
    public required IReadOnlyList<FlowStep> Steps { get; init; }
    /// <summary>True when the flow was auto-derived from the edges.</summary>
    public required bool Derived { get; init; }
}

/// <summary>A labelled horizontal band in a sequence diagram, drawn before message <c>At</c>.</summary>
internal sealed record SectionMark(string Label, int At, string Accent);

internal sealed record DiagramModel
{
    public required DiagramMeta Meta { get; init; }
    public required IReadOnlyList<NodeModel> Nodes { get; init; }
    public required IReadOnlyList<GroupModel> Groups { get; init; }
    public required IReadOnlyList<EdgeModel> Edges { get; init; }
    public required FlowModel Flow { get; init; }
    public required IReadOnlyList<SectionMark> Sections { get; init; }
}
